eth_defi/token_tax.py

```diff
@@ -0,0 +1,134 @@
+    :param base_token:
+        The token of which tax properties we are figuring out.
+
+    :param quote_token:
+        Address of the quote token used for the trading pair. E.g. `BUDS`, `WBNB`
+        Based on this information we can derive Uniswap trading pair address.
+
+    :param buy_account:
+        The account that does initial buy to measure the buy tax.
+        This account must be loaded with gas money (ETH/BNB) and `quote_token`
+        for a purchase.
+
+    :param sell_account:
+        The account that receives the token transfer and does the sell to measure the sell tax.
+        This account must be loaded with gas money for the sell.
+
+    :return:
+        ToxTaxInfo tells us what we figure out about taxes.
+        This can be later recorded to a database.
+    """
+
+    web3: Web3 = uniswap.web3
+    router = uniswap.router
+
+    quote_token_details = fetch_erc20_details(web3, quote_token)
+    quote_token = quote_token_details.contract
+
+    base_token_details = fetch_erc20_details(web3, base_token)
+    base_token = base_token_details.contract
+
+    # approve router to spend tokens
+    quote_token.functions.approve(router.address, quote_token_details.convert_to_raw(99999)).transact({"from": buy_account})
+
+    path = [quote_token.address, base_token.address]
+    amountIn = quote_token_details.convert_to_raw(1)
+    # Figure out base_token/quote_token trading pair
+    initial_base_bal = base_token.functions.balanceOf(buy_account).call()
+    # Buy base_token with buy_account
+    router.functions.swapExactTokensForTokensSupportingFeeOnTransferTokens(
+        amountIn,
+        0,
+        path,
+        buy_account,
+        FOREVER_DEADLINE
+    ).transact({"from": buy_account})
+
+    received_amt = base_token.functions.balanceOf(buy_account).call() - initial_base_bal
+    uniswap_price = router.functions.getAmountsOut(amountIn, path).call()[1]
+
+    # Measure the loss as "buy tax"
+    buy_tax_percent = (uniswap_price - received_amt) / uniswap_price
+
+    # Transfer tokens to sell_account
+    # Measure the loss as "transfer tax"
+    base_token.functions.transfer(sell_account, received_amt).transact({"from": buy_account})
+
+    received_amt_by_seller = base_token.functions.balanceOf(sell_account).call()
+
+    transfer_tax_percent = (received_amt - received_amt_by_seller) / received_amt
+
+    # Sell tokens
+    base_token.functions.approve(router.address, received_amt_by_seller).transact({"from": sell_account})
+    path = [ base_token.address, quote_token.address]
+
+    sell_tax = 0
+    sell_tax_percent = 0
+    try:
+        # this method will revert in case of low liquidity of the token
+        router.functions.swapExactTokensForTokensSupportingFeeOnTransferTokens(
+            received_amt_by_seller,
+            0,
+            path,
+            sell_account,
+            FOREVER_DEADLINE
+    ).transact({"from": sell_account})
+    except:
+        print("Low liquidity. Sell method failed")
```

Comment:
Create a test case for low liquidity path so that tests correctly check for a raised exception. See example here:

https://github.com/tradingstrategy-ai/web3-ethereum-defi/blob/e2d8265ee599bcb58fa62dddacf6899bcf2a9247/tests/test_token.py#L116

This can be spoofed by setting the the test buy amount to ridiculous high.